src/lib.ml

```diff
@@ -1306,6 +1306,24 @@ module DB = struct
       (virtual_lib, Variant.Map.of_list_exn [content]))
     |> Lib_name.Map.of_list_reduce ~f:Variant.Map.Multi.rev_union
 
+  (* implementations tagged with a variant are only variant when the they
+     implement a virtual library from the same project. *)
+  let check_valid_implementations (libmap : resolve_result Lib_name.Map.t) =
+    Lib_name.Map.iter libmap ~f:(function
+      | Found (lib : Lib_info.t) ->
+        begin match lib.implements, lib.variant with
+        | Some (loc, implements), Some variant ->
+          if not (Lib_name.Map.mem libmap implements) then
```

Comment:
This check looks a bit weak to me since `libmap` is for the whole workspace. We really need to check that the two libraries are part of the same project.

Reply:
Are you sure? This function is called for every project + its set of libraries.  I was under the impressions this is indeed per project and basically constructs the project's scope.

Reply:
Hmm, yh you are right. Could you add a test just to make sure we don't break this by accident?

Reply:
Sure, will do.

Reply:
Test case has been added